File: LinkedListPalindrom.java
//Linked List Palindrom - probably should have figued out a better way to find the mid point. 

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if(head == null) return true; 
        
        //Determine if its odd or even 
        ListNode temp = head;
        int size = 0;
        
        while(temp != null){
            temp = temp.next;
            size++;
        }
        
        // 1 -> 2 -> 1
        // length = 3
        // 3/2 = 1
      
        //mid +1 
        int mid = size/2;
            
        //If it is odd
        if(size % 2 != 0){
            ListNode oddMid = head;

            for(int i = 0; i <= mid; i++){
                oddMid = oddMid.next;
            }
            
            ListNode reversedRight = reverse(oddMid); 
            // 1 -> 1 -> 2 -> 1 -> 1
            ListNode leftSide = head;
            
            for(int i = 0; i < mid; i++){
                if(leftSide.val != reversedRight.val) return false;
                leftSide = leftSide.next;
                reversedRight = reversedRight.next;
            }
            
        }else{
            //If it is even
            ListNode evenMid = head;

            for(int i = 0; i < mid; i++){
                evenMid = evenMid.next;
            }
            
            ListNode reversedRight = reverse(evenMid); 
            // 1 -> 1 -> 2 -> 1 -> 1
            ListNode leftSide = head;
            
            for(int i = 0; i < mid; i++){
                if(leftSide.val != reversedRight.val) return false;
                leftSide = leftSide.next;
                reversedRight = reversedRight.next;
            }
        }
        
        return true;
        
    }
    
    public ListNode reverse(ListNode head){
        if(head == null || head.next == null){
            return head;
        } 

        ListNode rest = reverse(head.next);
        head.next.next = head;
        head.next = null;
        return rest;
    }
}

File: linkedListPalindrom.cpp
// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :


/*

In this problem, to find if the linked list is a palindrome or not, we first find the middle 
element by using fast and slow pointers. after find it , we then reverse the linked list start from the element after
the middle element and break the connection resulting in 2 linked lists.Then we will compare both linked lists and check 
if each node's value is equal or not. Whenever it encouters unequal values, it will
return false.


*/


// Your code here along with comments explaining your approach


#include <iostream>
#include <vector>
 using namespace std;


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {

        if(head->next == NULL){
            return true;
        }

        ListNode* slow = head;
        ListNode* fast = head;
        
        //find the mid

        while(fast->next != NULL && fast->next->next != NULL){
            slow = slow->next;
            fast = fast->next->next;
            

        }
         
         ListNode* prev = NULL;
         ListNode* curr = slow->next;
         ListNode* temp = curr->next;

         while(temp != NULL){
             curr->next = prev;
             prev = curr;
             curr = temp;
             temp = temp->next;
         }

         curr->next = prev;

         slow->next = NULL;

         fast = head;

         

         while(fast != NULL && curr != NULL){
             cout << fast->val << " " << curr->val ;
             if(fast->val==curr->val){
                
                fast= fast->next;
                curr=curr->next;


             }
             else{
                 
                 return false;
             }
         }

         return true;

    





        
    }
};